tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/TypedGradientAdapter.java

```diff
@@ -0,0 +1,88 @@
+/*
+ Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================
+*/
+package org.tensorflow.op;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.StatusFromTF_Status;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.GradientAdapterHelpers;
+import org.tensorflow.Graph;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.internal.c_api.GradFunc;
+import org.tensorflow.internal.c_api.NativeOperation;
+import org.tensorflow.internal.c_api.NativeOutputVector;
+import org.tensorflow.internal.c_api.NativeStatus;
+import org.tensorflow.internal.c_api.TF_Scope;
+import org.tensorflow.internal.c_api.TF_Status;
+
+/** A native adapter for {@link CustomGradient}. */
+public final class TypedGradientAdapter<T extends RawOp> extends GradFunc {
```

Comment:
I suggest to keep these `*GradientAdapter` package private as well, and maybe add static methods to `RawCustomGradient` and `CustomGradient`... or to move it to `org.tensorflow.internal` somewhere but it should definitely not be public as the extend directly from a Native class (`GradFunc`).

Might require moving some stuff out of `org.tensorflow.Tensorflow` as well to make everything well protected.